### ISSUE TITLE
[ERROR VoiceClient] Failed to start audio device.

### ISSUE THREAD

**Issue by geoffliddiard** [twilio/voice-callkit-quickstart-objc#9]
Hi,

I'm seeing this error intermittently when placing outgoing calls using SDK 2.0.0-beta6:

`2017-03-17 10:46:42.881322 AppName[7169:2427733] provider:performStartCallAction:`
`2017-03-17 10:46:42.954914 AppName[7169:2427733] provider:didActivateAudioSession:`
`2017-03-17 10:46:42.955050 AppName[7169:2427733] PJSIP(1):    pjsua_aud.c  .Error retrieving default audio device parameters: Invalid audio device (PJMEDIA_EAUD_INVDEV) [status=420004]`
`2017-03-17 10:46:42.955137 AppName[7169:2427733] [ERROR VoiceClient] Failed to start audio device.`

Sometimes audio still works, sometimes not. I've tried calling the configureAudioSession method outside of the provider inside `performStartCallActionWithUUID` but this hasn't made any difference. 

So far i've only tested this on iOS 10.2.1. Any advice or insight would be appreciated?

Thanks

**Comment by bchen-twilio** [comment]
Hi @geoffliddiard 

Bobie here from the SDK team and I would like to thank you for choosing the Programmable Voice SDK.
To be honest this is the first time I am seeing the `PJMEDIA_EAUD_INVDEV` error message and I might need more clues from you to triage the problem.
I would imagine this happened in your environment after making/answering a few calls, right? or were able to reproduce this problem with fresh launched app?

I'll also try if I can figure out any possible cause of this error.

-bobie

**Comment by geoffliddiard** [comment]
Hi @bchen-twilio 

Thanks for the response. I updated to the beta-7 sdk this morning and the error still persists. It now happens consistently when the app is launched and an outgoing call is placed for the first time. Subsequent outgoing calls are fine and audio works as expected (we only place outgoing calls at present). 

Here's the verbose output from that action if it helps:

```
2017-03-20 13:49:18.686537 ConsultantConnect[699:1151002] provider:performStartCallAction:
2017-03-20 13:49:18.686722 ConsultantConnect[699:1151002] [VERBOSE VoiceClient] Inside configureAudioSession, preparing audio session.
2017-03-20 13:49:18.689843 ConsultantConnect[699:1151002] [DEBUG TVOReachability] Reachability Flag Status: WR t------ networkStatusForFlags
2017-03-20 13:49:18.690221 ConsultantConnect[699:1151002] [DEBUG TVOCommandHandler] Inside postCommand:, received command of type TVOMakeCallCommand
2017-03-20 13:49:18.701540 ConsultantConnect[699:1151180] PJSIP(4): sip_endpoint.c !.Module "mod-pjsua-log" registered
2017-03-20 13:49:18.712196 ConsultantConnect[699:1151180] PJSIP(4):   pjsua_core.c  .Nameserver 208.67.222.222 added
2017-03-20 13:49:18.712950 ConsultantConnect[699:1151180] PJSIP(4): sip_endpoint.c  .Module "mod-tsx-layer" registered
2017-03-20 13:49:18.713131 ConsultantConnect[699:1151180] PJSIP(4): sip_endpoint.c  .Module "mod-stateful-util" registered
2017-03-20 13:49:18.713750 ConsultantConnect[699:1151180] PJSIP(4): sip_endpoint.c  .Module "mod-ua" registered
2017-03-20 13:49:18.715144 ConsultantConnect[699:1151180] PJSIP(4): sip_endpoint.c  .Module "mod-100rel" registered
2017-03-20 13:49:18.715658 ConsultantConnect[699:1151180] PJSIP(4): sip_endpoint.c  .Module "mod-pjsua" registered
2017-03-20 13:49:18.716667 ConsultantConnect[699:1151180] PJSIP(4): sip_endpoint.c  .Module "mod-invite" registered
2017-03-20 13:49:18.751403 ConsultantConnect[699:1151180] PJSIP(4): coreaudio_dev.  .. dev_id 0: iPhone IO device  (in=1, out=1) 8000Hz
2017-03-20 13:49:18.766676 ConsultantConnect[699:1151002] provider:didActivateAudioSession:
2017-03-20 13:49:18.766862 ConsultantConnect[699:1151002] [VERBOSE VoiceClient] Inside startAudioDevice, starting audio device.
2017-03-20 13:49:18.767009 ConsultantConnect[699:1151002] PJSIP(4):    pjsua_aud.c !Set sound device: capture=-1, playback=-2
2017-03-20 13:49:18.767105 ConsultantConnect[699:1151002] PJSIP(1):    pjsua_aud.c  .Error retrieving default audio device parameters: Invalid audio device (PJMEDIA_EAUD_INVDEV) [status=420004]
2017-03-20 13:49:18.767210 ConsultantConnect[699:1151002] [ERROR VoiceClient] Failed to start audio device.
2017-03-20 13:49:18.776792 ConsultantConnect[699:1151180] PJSIP(4): coreaudio_dev. !..core audio initialized
2017-03-20 13:49:18.795407 ConsultantConnect[699:1151180] PJSIP(4):          pjlib  ..select() I/O Queue created (0x10db0c028)
2017-03-20 13:49:18.797662 ConsultantConnect[699:1151180] PJSIP(5):   conference.c  ..Creating conference bridge with 40 ports
2017-03-20 13:49:18.797789 ConsultantConnect[699:1151180] PJSIP(5):   Master/sound  ..Using delay buffer with WSOLA.
WARNING: no real random source present!


```

Thanks

Geoff

**Comment by bchen-twilio** [comment]
Thanks for the log and the experiment analysis Geoff!
Unfortunately I still can't tell what exactly is going wrong but I've created an internal JIRA ticket for tracking this.
Could you tell me if there is any `AVAudioSession` related manipulation code in your app? for example, dial-tone or ringtone playback.

-bobie

**Comment by bchen-twilio** [comment]
Hi @geoffliddiard 

How's it going? I've tried some experiments and was able to somehow reproduce the `PJMEDIA_EAUD_INVDEV` error. Here's what I tried:
- Upon incoming call invite, I use the AVAudioPlayer to play some incoming call ringtone
- CallKit integration enabled in my app
- Before playing I change the AVAudioSession category to `AVAudioSessionCategoryPlayback`

Usually before calling `TVOCallInvite.accept`, my app will first stop the ringtone playback and set the category back to `AVAudioSessionCategoryPlayAndRecord`. I found that if CallKit is enabled in the app, the app logic won't restore the category and I am seeing the same failure as you are.

Could you try to make sure you have the correct category (`.playAndRecord`) before calling `.accept`?
Thanks.

**Comment by geoffliddiard** [comment]
Hey @bchen-twilio,

Apologies for the slow response and thanks for your efforts so far. We haven't implemented incoming calls in our app yet, so we're only seeing this issue when placing outgoing calls. We have a speaker button which sets the category `AVAudioSessionCategoryPlayAndRecord` and toggles routing to the device speaker, other than that we're not doing any audio stuff. For the time being we've reverted to SDK v2.0.0-beta5 which is stable for our use case.

We could knock together a demo app with steps to reproduce if you'd be willing to take a further look at it?

Cheers

**Comment by bchen-twilio** [comment]
Hi @geoffliddiard 

Could you try using these lines to route your call audio?
```
  // speaker
  [[AVAudioSession sharedInstance] overrideOutputAudioPort:kAudioSessionOverrideAudioRoute_Speaker error:nil];
  // default device
  [[AVAudioSession sharedInstance] overrideOutputAudioPort:kAudioSessionOverrideAudioRoute_None error:nil];
```

This is what we are using in our internal testbed app and it's working great for us. We are sorry that the audio session configure is quite delicate and does not cooperate sometimes.
Thanks.

Also we've recently found an audio device setup bug when doing integration with CallKit. It'd be great if you could give this patch a try and let us know if it works for you too: https://drive.google.com/file/d/0B0jCXbOI9plrczFDZV91MFFWeE0/view?usp=sharing

A couple of SDK users and our QE team found that sometime when launching the app and make the first outgoing call, the call audio doesn't really work. We did some call flow review and refactored a bit hopefully to make the CallKit integration better. Funny thing is that this doesn't happen (or rarely happens) when running the app with Xcode debugger instead of running without the debugger or running unplugged. 

Cheers,
-bobie

**Comment by geoffliddiard** [comment]
Hey @bchen-twilio 

The patch works great! The problem sounds almost identical to our experience, we would see it happening in about 1 in 10 builds running on the debugger and more frequently unplugged as you mentioned. No longer an issue using the patched version.

Cheers

**Comment by bchen-twilio** [comment]
@geoffliddiard 
Sounds great!
We should be releasing the next beta to include this bugfix within one or two iterations of RCs. Please stay tuned for updates.

Thanks.
Btw I removed the link to the patch.

-bobie

**Comment by bchen-twilio** [comment]
Hi @geoffliddiard 

We've just released the Voice iOS SDK 2.0.0-beta8. Please update your application with the latest SDK and please note that we've changed the name of the framework and the face object to both `TwilioVoice`, in case you might encounter some conflicts during update. You will also have to update the framework name in your `Podfile` (if you are using Cocoapods).

We've recently combined regular quickstart and CallKit quickstart into the same repo (https://github.com/twilio/voice-quickstart-objc) and we are going to close this repo soon. I am going to close this ticket for now, but please feel free to open a new issue in these two repositories:
https://github.com/twilio/voice-quickstart-objc
https://github.com/twilio/voice-quickstart-swift

Thanks again for choosing the Twilio Voice SDK.

-bobie
